components/fonts/font_cache_thread.rs

```diff
@@ -316,20 +338,29 @@ pub trait FontSource: Clone {
         size: Au,
         flags: FontInstanceFlags,
     ) -> FontInstanceKey;
-    fn get_web_font(&self, data: Arc<Vec<u8>>, index: u32) -> FontKey;
+    fn get_web_font(&self, data: Arc<FontData>, index: u32) -> FontKey;
     fn get_web_font_instance(
         &self,
         font_key: FontKey,
         size: f32,
         flags: FontInstanceFlags,
     ) -> FontInstanceKey;
+    fn get_font_data(&self, identifier: &FontIdentifier) -> Option<Arc<FontData>>;
+}
+
+#[derive(Debug, Eq, Hash, MallocSizeOf, PartialEq)]
+struct FontTemplateCacheKey {
+    font_descriptor: Option<FontDescriptor>,
+    family_descriptor: SingleFontFamily,
 }
 
 /// The public interface to the font cache thread, used by per-thread `FontContext` instances (via
 /// the `FontSource` trait), and also by layout.
-#[derive(Clone, Debug, Deserialize, Serialize)]
-pub struct FontCacheThread {
-    chan: IpcSender<Command>,
+#[derive(Debug)]
+pub struct SystemFontServiceProxy {
+    sender: ReentrantMutex<IpcSender<Command>>,
+    templates: RwLock<HashMap<FontTemplateCacheKey, Vec<FontTemplateRef>>>,
+    data_cache: RwLock<HashMap<FontIdentifier, Arc<FontData>>>,
```

Comment:
I would move the arc to inside here, and drop the mutex around the sender(unless somehow necessary), and remove the arc around the proxy elsewhere.

Reply:
This is necessary because the proxy needs shared ownership and interior mutability. It's owned by each per-Document `FontContext` and represents the shared cache for all threads running in the content process.

Reply:
That said, if we can verify that only a single thread in the content process is using the content process' `SystemFontServiceProxy`, then we could probably convert this into something behind an `Rc` instead of an `Arc`.

Reply:
Additionally, I'll note here that putting the `RwLock` around each member is meant to reduce contention if two threads are trying to query system fonts at once.

Reply:
Ok, perhaps good for a follow-up investigation then.